### PR TITLE
Support backup encryption

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Signed-off-by: Yi Wu <yiwu@pingcap.com>

### What problem does this PR solve? 
Support encrypting a backup. User can specify `--encryption` and `--encryption-key` in BR command line to encryption a backup, and restore a encrypted backup. Both of backup meta and SST files gets encrypted.

### What is changed and how it works?
* command-line encryption arguments are parsed and used for both encrypting/decrypting `BackupMeta` and pass to tikv in `BackupRequest` and `DownloadRequest`.
* despite command-line argument specifies CTR mode as encryption method, `BackupMeta` is actually encrypted with GCM mode for extra authentication.

### Check List 
Not tested yet. Will update.

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
WIP: code completed but not tested yet.

Completely newbie of golang and BR. Please review carefully. :)

File: go.mod [hunk 1]
```diff
@@ -31,14 +32,18 @@ require (
...
+	github.com/twitchtv/retool v1.3.7 // indirect
 	go.etcd.io/etcd v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738
 	go.opencensus.io v0.22.2 // indirect
 	go.uber.org/zap v1.14.1
 	golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45
 	google.golang.org/api v0.14.0
 	google.golang.org/grpc v1.25.1
 )
+
+replace github.com/pingcap/kvproto => github.com/yiwu-arbug/kvproto v0.0.0-20200331222645-f72d7b6bf777
```
Comment (yiwu-arbug) [inline go.mod:49]: need to remove after merging https://github.com/pingcap/kvproto/pull/592

File: pkg/encryption/flags.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	flagEncryptionKey = "encryption-key"
+)
+
+// DefineFlags adds encryption related flags.
+func DefineFlags(flags *pflag.FlagSet) {
+	flags.String(flagEncryption, "", "Enable encryption and use the specific"+
+		" method if non-empty. Can be one of aes128-ctr, aes192-ctr oraes256-ctr.")
+	// The prefix of the key is to support key management service later.
+	// For example to specify a KMS CMK, we can allow 'kms:<cmk-key-id>' in the future.
+	flags.String(flagEncryptionKey, "", "Encryption key to be use, if encryption"+
```
Comment (yiwu-arbug) [inline pkg/encryption/flags.go:26]: need to change to pass key from file instead of directly from command-line.

**Comment by yiwu-arbug** [comment]
ready for review. will add more test of PR look okay.

File: pkg/encryption/common.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	Config         encryptionpb.EncryptionConfig
+	keyBackendType string
+}
+
+type keyBackend interface {
+	GetKey() (key []byte, encryptedKey []byte, err error)
+	DecryptKey(encryptedKey []byte) (key []byte, err error)
+}
+
+func (options *Options) fillDataKey() error {
```
Comment (yiwu-arbug) [inline pkg/encryption/common.go:42]: According to discussion on the docs, will change this to derive the key from backupmeta key using HKDF: https://docs.google.com/document/d/1x5xx8Zbm-frzj1-t5OnMcpEUQHjpnl6vSkCmaehpeLM/edit?disco=AAAAJZ6wItk

Comment (kennytm) [reply pkg/encryption/common.go:42]: (note that the Google doc is not public. we should move these design documents into this repo in the future.)

**Comment by yiwu-arbug** [comment]
We will be using S3 server-side encryption to support BR encryption instead. Not moving forward for this one for now.

**Pull request closed without merging**